## Gather

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [3]:
patients

patient_id assigned_sex  given_name       surname  \
0             1       female         Zoe       Wellish   
1             2       female      Pamela          Hill   
2             3         male         Jae        Debord   
3             4         male        Liêm          Phan   
4             5         male         Tim       Neudorf   
5             6         male      Rafael         Costa   
6             7       female        Mary         Adams   
7             8       female      Xiuxiu         Chang   
8             9         male       Dsvid    Gustafsson   
9            10       female      Sophie       Cabrera   
10           11       female       Sandy    Gunnarsson   
11           12         male   Abdul-Nur           Isa   
12           13         male  Omeokachie      Ibeamaka   
13           14       female    Anenechi         Chidi   
14           15       female        Asia       Woźniak   
15           16         male       Søren          Lund   
16           17       female         Tám          Liễu   
17           18       female     Roxanne     Andreyeva   
18           19         male     William         Oates   
19           20         male         Zak         Kelly   
20           21       female       Sofia       Karlsen   
21           22         male      Samúel  Guðbrandsson   
22           23         male      Manchu            Su   
23           24         male       Lovre         Galić   
24           25         male       Jakob      Jakobsen   
25           26         male      Gregor          Bole   
26           27       female        Ella          Lund   
27           28         male      Joseph        Tucker   
28           29         male      Robert          Wolf   
29           30         male        Jake      Jakobsen   
..          ...          ...         ...           ...   
473         474       female        Kate     Wilkinson   
474         475       female   Esperanza      Labrosse   
475         476         male       Malik       Vaneker   
476         477       female       Berta    Napolitani   
477         478         male     Juliusz      Majewski   
478         479       female      Edelma   Villalpando   
479         480         male        Tapa   Arsanukayev   
480         481         male      Nasser       Mansour   
481         482         male     Michael    Kristensen   
482         483         male       Diogo         Souza   
483         484       female       Angel         Grant   
484         485         male     Placido       Udinesi   
485         486         male      Trifon      Izmailov   
486         487         male      Samuel          Blix   
487         488         male        Ivar       Löfgren   
488         489         male        Mika    Martinsson   
489         490       female     Jasmine         Sykes   
490         491         male     Jackson       Addison   
491         492       female     Vanessa      Ferguson   
492         493         male       Poldi           Tar   
493         494       female         Fen          Chin   
494         495       female      Sirkka     Piirainen   
495         496         male      Hajime       Tsukada   
496         497         male   Alexander        Hueber   
497         498         male    Masataka      Murakami   
498         499         male     Mustafa     Lindström   
499         500         male       Ruman       Bisliev   
500         501       female       Jinke     de Keizer   
501         502       female     Chidalu   Onyekaozulu   
502         503         male         Pat       Gersten   

                        address               city       state  zip_code  \
0          576 Brown Bear Drive  Rancho California  California   92390.0   
1     2370 University Hill Road          Armstrong    Illinois   61812.0   
2         1493 Poling Farm Road               York    Nebraska   68467.0   
3           2335 Webster Street         Woodbridge          NJ    7095.0   
4          1428 Turkey Pen L

In [4]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0       veronika      jindrová  41u - 48u          -         7.63       7.20   
1         elliot    richardson          -  40u - 45u         7.56       7.09   
2       yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3           skye    gormanston  33u - 36u          -         7.97       7.62   
4         alissa        montez          -  33u - 29u         7.78       7.46   
5        jasmine         sykes          -  42u - 44u         7.56       7.18   
6         sophia        haugen  37u - 42u          -         7.65       7.27   
7          eddie        archer  31u - 38u          -         7.89       7.55   
8          saber        ménard          -  54u - 54u         8.08       7.70   
9           asia       woźniak  30u - 36u          -         7.76       7.37   
10        joseph           day  29u - 36u          -         7.70       7.19   
11     kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12       roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13        gregor          bole          -  47u - 45u         7.61       7.16   
14        simone  baumgaertner  27u - 37u          -         7.74       7.30   
15          enco        žibrik  55u - 68u          -         7.78       7.34   
16       camilla      zaitseva  28u - 37u          -         7.53       7.13   
17          gina          cain          -  36u - 36u         7.88       7.40   
18    addolorata      lombardi          -  49u - 46u         7.75       7.33   
19        khalid      johnsrud          -  54u - 54u         8.35       7.94   
20          mile        stanić          -  47u - 48u         7.66       7.24   
21         tekla       walczak  29u - 39u          -         7.61       7.29   
22   brancaleone         russo  53u - 60u          -         8.61       8.18   
23      chiemela     tobeolisa          -  43u - 47u         7.59       7.17   
24          isac          berg  31u - 41u          -         9.68       9.29   
25        benoît        bonami          -  44u - 43u         9.82       9.40   
26        suhaim         rahal          -  49u - 47u         7.94       7.50   
27        mizuki         iwata          -  45u - 46u         7.70       7.23   
28       clinton        miller  42u - 51u          -         7.79       7.40   
29        eugene       mironov  42u - 49u          -         7.81       7.48   
..           ...           ...        ...        ...          ...        ...   
250         chen           yao          -  56u - 57u         7.90       7.51   
251        aksel   vestergaard          -  42u - 38u         9.62       9.29   
252        ellen         luman          -  40u - 39u         9.27       8.77   
253       albino     schiavone  35u - 43u          -         7.56       7.15   
254         jose         combs          -  39u - 36u         7.89       7.42   
255       jia li          teng  48u - 54u          -         7.66       7.32   
256        ilija        horvat  42u - 50u          -         7.77       7.38   
257     mathilde      nørgaard          -  27u - 28u         8.50       8.10   
258       csilla      herczegh          -  43u - 46u         7.71       7.27   
259      aaliyah          rice          -  31u - 31u         7.64       7.33   
260        david      beauvais          -  26u - 23u         7.87       7.47   
261     caroline        shuler          -  50u - 54u         7.63       7.27   
262         alex      crawford  51u - 62u          -         7.69       7.30   
263      rebecca      jephcott  53u - 63u          -         7.96       7.57   
264   chukwumoge     ogochukwu          -  41u - 39u         7.95       7.56   
265       fearne      mcgregor          -  27u - 29u         7.83       7.48   
266       ursula         freud  42u - 54u          -         7.75       7.46   
267         leon        scholz          -  38u - 32u         7.72       7.29   
268   

In [5]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [8]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [9]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [10]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [11]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [12]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [13]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [14]:
patients.sample(5)

patient_id assigned_sex given_name    surname                  address  \
258         259         male       Abel    Yonatan         2621 Koontz Lane   
121         122         male    Frydryk    Adamski  1845 Saint Marys Avenue   
122         123         male        Bội         Tạ       2389 Rubaiyat Road   
111         112       female     Nicole  Zimmerman      909 Williams Avenue   
353         354         male        Ole   Petersen     3650 Graystone Lakes   

             city       state  zip_code        country  \
258       Burbank  California   91502.0  United States   
121      Syracuse    New York   13221.0  United States   
122  Grand Rapids          MI   49503.0  United States   
111       Newhall  California   91321.0  United States   
353        Dudley          GA   31022.0  United States   

                                      contact   birthdate  weight  height  \
258     AbelYonatan@teleworm.us1 818 841 7660   4/29/1952   137.9      66   
121  FrydrykAdamski@armyspy.com1 315 869 4408  10/12/1941   230.8      76   
122         TaHaBoi@superrito.com231-607-3625   5/30/1929   211.0      69   
111     661-291-1812NicoleZimmerman@rhyta.com   1/14/1984   225.9      74   
353       OlePetersen@armyspy.com478-676-7058   7/14/1931   136.8      69   

      bmi  
258  22.3  
121  28.1  
122  31.2  
111  29.0  
353  20.2

In [15]:
patients.surname.value_counts()

Doe             6
Jakobsen        3
Taylor          3
Kadyrov         2
Correia         2
Tucker          2
Silva           2
Tạ              2
Gersten         2
Grímsdóttir     2
Ogochukwu       2
Schiavone       2
Cindrić         2
Lund            2
Souza           2
Berg            2
Bùi             2
Lương           2
Liễu            2
Lâm             2
Dratchev        2
Hueber          2
Kowalczyk       2
Johnson         2
Collins         2
Batukayev       2
Aranda          2
Cabrera         2
Nilsen          2
Parker          2
               ..
Lindgren        1
Gormanston      1
Adams           1
Haraguchi       1
Richardson      1
Mathiesen       1
Day             1
Mata            1
Štěpánek        1
Aličajić        1
Wiśniewska      1
Rokavc          1
Kašná           1
Azevedo         1
Ščančar         1
Efrem           1
Thạch           1
Pak             1
Nnonso          1
Jakšić          1
Stanić          1
House           1
Frederiksen     1
Guðbrandsson    1
Reynoso   

In [16]:
patients.address.value_counts()

123 Main Street               6
648 Old Dear Lane             2
2476 Fulton Street            2
2778 North Avenue             2
3268 Karen Lane               1
4310 Johnson Street           1
3259 Roy Alley                1
1384 Dovetail Estates         1
2102 Geraldine Lane           1
475 Preston Street            1
4992 Moonlight Drive          1
1270 Haul Road                1
936 Lightning Point Drive     1
4019 Cerullo Road             1
4102 Ritter Avenue            1
163 Hide A Way Road           1
1389 Nutters Barn Lane        1
3479 Elm Drive                1
479 Elmwood Avenue            1
4970 Heather Sees Way         1
435 Pike Street               1
2691 Kessla Way               1
4237 Hamilton Drive           1
1953 Rhapsody Street          1
1904 Granville Lane           1
883 Oakwood Circle            1
4900 Philli Lane              1
272 Boone Crockett Lane       1
1147 Kelley Road              1
796 Eagle Street              1
                             ..
2043 Jad

In [17]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [18]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [19]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [20]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [21]:
sum(treatments.auralin.isnull())

0

In [22]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [23]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

<font color='red'>Complete the following two "Missing Data" **Define, Code, and Test** sequences after watching the *"Address Missing Data First"* video.</font>


#### `treatments`: Missing records (280 instead of 350)


*Your definition here. Note: the missing `treatments` records are stored in a file named `treatments_cut.csv`, which you can see in this Jupyter Notebook's dashboard (click the **jupyter** logo in the top lefthand corner of this Notebook). Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) for the function used in the solution.*

##### Define
Add the missing records to treatments dataframe by using pd.append() 

##### Code

In [24]:
# Your cleaning code here
treatments_cut = pd.read_csv('treatments_cut.csv')
#treatments_cut.head()
treatments_clean = treatments_clean.append(treatments_cut)  


##### Test

In [25]:
# Your testing code here
# treatments_complete.shape
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 69
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    213 non-null float64
dtypes: float64(3), object(4)
memory usage: 21.9+ KB


#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

##### Define
*Your definition here.*
recalculate all the HbA1c changes by subtracting hba1c_end with hba1c_start

##### Code

In [26]:
# Your cleaning code here
treatments_clean['hba1c_change'] = treatments_clean['hba1c_start'] - treatments_clean['hba1c_end']

##### Test

In [27]:
# Your testing code here
treatments_clean['hba1c_change']

0     0.43
1     0.47
2     0.43
3     0.35
4     0.32
5     0.38
6     0.38
7     0.34
8     0.38
9     0.39
10    0.51
11    0.38
12    0.40
13    0.45
14    0.44
15    0.44
16    0.40
17    0.48
18    0.42
19    0.41
20    0.42
21    0.32
22    0.43
23    0.42
24    0.39
25    0.42
26    0.44
27    0.47
28    0.39
29    0.33
      ... 
40    0.28
41    0.46
42    0.47
43    0.40
44    0.32
45    0.47
46    0.29
47    0.51
48    0.39
49    0.38
50    0.40
51    0.29
52    0.43
53    0.37
54    0.37
55    0.39
56    0.44
57    0.37
58    0.45
59    0.30
60    0.37
61    0.36
62    0.44
63    0.39
64    0.41
65    0.34
66    0.45
67    0.30
68    0.47
69    0.46
Name: hba1c_change, Length: 350, dtype: float64

### Tidiness

<font color='red'>Complete the following four "Tidiness" **Define, Code, and Test** sequences after watching the *"Cleaning for Tidiness"* video.</font>

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
*Your definition here. Hint 1: use regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). Here is an amazing [regex tutorial](https://regexone.com/). Hint 2: [various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number). Hint 3: [email address regex pattern](http://emailregex.com/), which you might need to modify to distinguish the email from the phone number.*

split contact column in patients table into phone number and email two columns by using regex and drop the contact column

##### Code

In [28]:
# Your cleaning code here
patients_clean['phone_number'] = patients_clean['contact'].str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)
patients_clean['email'] = patients_clean['contact'].str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])',expand=True)

In [29]:
# Note: axis=1 denotes that we are referring to a column, not a row
patients_clean = patients_clean.drop('contact', axis=1)

##### Test

In [30]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  birthdate  weight  \
0  Rancho California  California   92390.0  United States  7/10/1976   121.7   
1          Armstrong    Illinois   61812.0  United States   4/3/1967   118.8   
2               York    Nebraska   68467.0  United States  2/19/1980   177.8   
3         Woodbridge          NJ    7095.0  United States  7/26/1951   220.9   
4             Dothan          AL   36303.0  United States  2/18/1928   192.3   

   height   bmi       phone_number                      email  
0      66  19.6       951-719-9170   ZoeWellish@superrito.com  
1      66  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2      71  24.8       402-363-6804       JaeMDebord@gustr.com  
3      70  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4      27  26.1       334-515-7487        TimNeudorf@cuvox.de

In [31]:
# Your testing code here
patients_clean['phone_number'].head()

0         951-719-9170
1    +1 (217) 569-3204
2         402-363-6804
3    +1 (732) 636-8246
4         334-515-7487
Name: phone_number, dtype: object

In [32]:
patients_clean['phone_number'].tail()

498    207-477-0579
499    928-284-4492
500    816-223-6007
501    360 443 2060
502    402-848-4923
Name: phone_number, dtype: object

In [33]:
patients_clean['email'].head()

0     ZoeWellish@superrito.com
1         PamelaSHill@cuvox.de
2         JaeMDebord@gustr.com
3    PhanBaLiem@jourrapide.com
4          TimNeudorf@cuvox.de
Name: email, dtype: object

In [34]:
patients_clean['email'].tail()

498      MustafaLindstrom@jourrapide.com
499               RumanBisliev@gustr.com
500            JinkedeKeizer@teleworm.us
501    ChidaluOnyekaozulu@jourrapide.com
502             PatrickGersten@rhyta.com
Name: email, dtype: object

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
*Your definition here. Hint: use pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). Here is an excellent [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

split the start_dose and end_dose by "-" , start_dose as a column , end_dose as a column.  Melt start_dose and end_dose into 3 columns with pd.melt() 

##### Code

In [35]:
# Your cleaning code here
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

In [36]:
# melt two columns(auralin and novodra) into one(treatment)
# all the rest columns can be id_vals, mutiple val ids so put them into a list
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')
treatments_clean = treatments_clean[treatments_clean.dose != "-"]
treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str
treatments_clean = treatments_clean.drop('dose', axis=1)


##### Test

In [37]:
# Your testing code here
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  dose_start dose_end  
0        41u      48u  
3        33u      36u  
6        37u      42u  
7        31u      38u  
9        30u      36u

#### Adverse reaction should be part of the `treatments` table

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for the function used in the solution.*

Merge the adverse_reaction table into treatment table joining on given name and surname

##### Code

In [38]:
# Your cleaning code here
treatments_clean = pd.merge(treatments_clean , adverse_reactions_clean , on= ['surname' ,'given_name' ], how='left')
# multiple conditions : use list 
# based on patient table, so left join 

##### Test

In [39]:
# Your testing code here
treatments_clean.sample(5)

given_name    surname  hba1c_start  hba1c_end  hba1c_change treatment  \
36       julia     werner         7.58       7.24          0.34   auralin   
16   alexander  mathiesen         7.96       7.55          0.41   auralin   
70     manuela    cindrić         8.07       7.76          0.31   auralin   
210       tosh     jensen         7.93       7.69          0.24   novodra   
126      ilija     horvat         7.77       7.38          0.39   auralin   

    dose_start dose_end adverse_reaction  
36         35u      40u              NaN  
16         47u      58u              NaN  
70         55u      66u              NaN  
210        51u      48u              NaN  
126        42u      50u              NaN

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution.*

Replace the given_name and surname columns in treatments_clean with patient_ID 
1)create a new dataframe "names_id" with the surname, given_name and patient ID from patients table, 
2) change surnames and given_names to lower case 
3)merge names_id with treatment table on surname and given_name

##### Code

In [40]:
# Your cleaning code here
id_names = patients_clean[['patient_id', 'given_name', 'surname']]
id_names.given_name = id_names.given_name.str.lower()
id_names.surname = id_names.surname.str.lower()
treatments_clean = pd.merge(treatments_clean, id_names, on=['given_name', 'surname'])
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)


/opt/conda/lib/python3.6/site-packages/pandas/core/generic.py:3110: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


##### Test

In [41]:
# Your testing code here
treatments_clean

hba1c_start  hba1c_end  hba1c_change treatment dose_start dose_end  \
0           7.63       7.20          0.43   auralin        41u      48u   
1           7.97       7.62          0.35   auralin        33u      36u   
2           7.65       7.27          0.38   auralin        37u      42u   
3           7.89       7.55          0.34   auralin        31u      38u   
4           7.76       7.37          0.39   auralin        30u      36u   
5           7.70       7.19          0.51   auralin        29u      36u   
6           7.70       7.19          0.51   auralin        29u      36u   
7           9.54       9.14          0.40   auralin        29u      38u   
8           7.74       7.30          0.44   auralin        27u      37u   
9           7.78       7.34          0.44   auralin        55u      68u   
10          7.53       7.13          0.40   auralin        28u      37u   
11          7.61       7.29          0.32   auralin        29u      39u   
12          8.61       8.18          0.43   auralin        53u      60u   
13          9.68       9.29          0.39   auralin        31u      41u   
14          7.79       7.40          0.39   auralin        42u      51u   
15          7.81       7.48          0.33   auralin        42u      49u   
16          7.70       7.38          0.32   auralin        35u      39u   
17          7.96       7.55          0.41   auralin        47u      58u   
18          7.68       7.24          0.44   auralin        45u      48u   
19          7.92       7.47          0.45   auralin        24u      37u   
20          7.92       7.57          0.35   auralin        44u      55u   
21          7.53       7.15          0.38   auralin        37u      43u   
22          7.67       7.37          0.30   auralin        43u      47u   
23          7.86       7.51          0.35   auralin        36u      42u   
24          9.18       8.64          0.54   auralin        29u      37u   
25          7.61       7.16          0.45   auralin        57u      64u   
26          7.52       7.11          0.41   auralin        54u      67u   
27          7.71       7.34          0.37   auralin        34u      42u   
28          7.87       7.47          0.40   auralin        29u      37u   
29          7.74       7.32          0.42   auralin        61u      67u   
..           ...        ...           ...       ...        ...      ...   
319         7.98       7.60          0.38   novodra        37u      38u   
320         7.61       7.14          0.47   novodra        28u      24u   
321         7.99       7.56          0.43   novodra        24u      26u   
322         7.73       7.34          0.39   novodra        20u      21u   
323         7.71       7.30          0.41   novodra        46u      46u   
324         7.54       7.17          0.37   novodra        26u      27u   
325         7.51       7.08          0.43   novodra        34u      33u   
326         7.89       7.55          0.34   novodra        30u      35u   
327         7.71       7.30          0.41   novodra        33u      33u   
328         8.00       7.50          0.50   novodra        35u      34u   
329         7.79       7.45          0.34   novodra        30u      27u   
330         7.70       7.36          0.34   novodra        37u      41u   
331         7.84       7.37          0.47   novodra        43u      40u   
332         8.90       8.57          0.33   novodra        34u      34u   
333         7.64       7.24          0.40   novodra        40u      36u   
334         7.54       7.26          0.28   novodra        37u      40u   
335         7.67       7.21          0.46   novodra        26u      23u   
336         9.41       8.94          0.47   novodra        40u      45u   
337         7.68       7.36          0.32   novodra        52u      52u   
338         7.78       7.39          0.39   novodra        37u      39u   
339         7.63       7.25          0.38   novodra        42u      42u   
340         7.92       7.52   

In [42]:
# Patient ID should be the only duplicate column
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

22    patient_id
dtype: object

### Quality

<font color='red'>Complete the remaining "Quality" **Define, Code, and Test** sequences after watching the *"Cleaning for Quality"* video.</font>

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
*Your definition here. Hint: see the "Data Cleaning Process" page.*
* convert float to string using
* remove .0
* pad leading 0 to make sure each str zip_code is 5 

##### Code

In [43]:
sum(patients['zip_code'].isnull())

12

In [44]:
# Your cleaning code here
patients['zip_code'] = patients['zip_code'].astype(str).str[:-2] #.str.rjust(5,'0') # str[:-2] to get rid of decimal


In [45]:
import numpy as np
# memthod 1:
#for i in range(len(patients['zip_code'])): # len(patients['zip_code'] refers to all the rows under this column
     #print(patients['zip_code'][i])
#    if len(patients['zip_code'][i])== 4:
#        patients['zip_code'][i] = '0'+ patients['zip_code'][i]
        
# method 2: this method is better coz there may be 3-digit zip code cases
patients_clean['zip_code'] = patients_clean['zip_code'].astype(str).str[:-2].str.rjust(5,'0') 
# str[:-2] to get rid of decimal
# rjust(5,'0') to pad 0 at the left side
# nan should be condsider too
patients_clean['zip_code'] = patients_clean['zip_code'].replace('0000n', np.nan)


##### Test

In [46]:
# Your testing code here
patients_clean['zip_code'].head()

0    92390
1    61812
2    68467
3    07095
4    36303
Name: zip_code, dtype: object

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
*Your definition here.*
replace the value 27 in height column with 72

##### Code

In [47]:
# Your cleaning code here
patients_clean.height = patients_clean.height.replace(27,72)


##### Test

In [48]:
# Your testing code here
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state zip_code        country  birthdate  weight  \
0  Rancho California  California    92390  United States  7/10/1976   121.7   
1          Armstrong    Illinois    61812  United States   4/3/1967   118.8   
2               York    Nebraska    68467  United States  2/19/1980   177.8   
3         Woodbridge          NJ    07095  United States  7/26/1951   220.9   
4             Dothan          AL    36303  United States  2/18/1928   192.3   

   height   bmi       phone_number                      email  
0      66  19.6       951-719-9170   ZoeWellish@superrito.com  
1      66  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2      71  24.8       402-363-6804       JaeMDebord@gustr.com  
3      70  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4      72  26.1       334-515-7487        TimNeudorf@cuvox.de

#### Full state names sometimes, abbreviations other times

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) for method used in solution.*
* create a function to change all the full name states to abbreviation with lambda


##### Code

In [49]:
# Your cleaning code here
us_state_abbrev = {
    'Alabama': 'AL',
    'Alaska': 'AK',
    'Arizona': 'AZ',
    'Arkansas': 'AR',
    'California': 'CA',
    'Colorado': 'CO',
    'Connecticut': 'CT',
    'Delaware': 'DE',
    'Florida': 'FL',
    'Georgia': 'GA',
    'Hawaii': 'HI',
    'Idaho': 'ID',
    'Illinois': 'IL',
    'Indiana': 'IN',
    'Iowa': 'IA',
    'Kansas': 'KS',
    'Kentucky': 'KY',
    'Louisiana': 'LA',
    'Maine': 'ME',
    'Maryland': 'MD',
    'Massachusetts': 'MA',
    'Michigan': 'MI',
    'Minnesota': 'MN',
    'Mississippi': 'MS',
    'Missouri': 'MO',
    'Montana': 'MT',
    'Nebraska': 'NE',
    'Nevada': 'NV',
    'New Hampshire': 'NH',
    'New Jersey': 'NJ',
    'New Mexico': 'NM',
    'New York': 'NY',
    'North Carolina': 'NC',
    'North Dakota': 'ND',
    'Ohio': 'OH',
    'Oklahoma': 'OK',
    'Oregon': 'OR',
    'Pennsylvania': 'PA',
    'Rhode Island': 'RI',
    'South Carolina': 'SC',
    'South Dakota': 'SD',
    'Tennessee': 'TN',
    'Texas': 'TX',
    'Utah': 'UT',
    'Vermont': 'VT',
    'Virginia': 'VA',
    'Washington': 'WA',
    'West Virginia': 'WV',
    'Wisconsin': 'WI',
    'Wyoming': 'WY',
}

# the parameter in the function should be the same level with the varible to be used later
# patients is the same level with patients_clean 
def abbrevi(patients):
    if patients['state'] in us_state_abbrev.keys():
        abbre = us_state_abbrev[patients['state']]
        return abbre
    #else:
    if patients['state'] in us_state_abbrev.values():
        return patients['state']
        
patients_clean['state'] = patients_clean.apply(abbrevi,axis=1)



##### Test

In [50]:
# Your testing code here
patients_clean['state']

0      CA
1      IL
2      NE
3      NJ
4      AL
5      FL
6      NV
7      CA
8      MO
9      NY
10     MI
11     TN
12     VA
13     AL
14     OK
15     VA
16     CA
17     GA
18     MT
19     CA
20     MA
21     NJ
22     NY
23     FL
24     NY
25     NM
26     IL
27     LA
28     PA
29     NY
       ..
473    MA
474    GA
475    CA
476    PA
477    SC
478    CA
479    CA
480    TX
481    OK
482    TX
483    TN
484    NC
485    FL
486    TX
487    KS
488    FL
489    CA
490    NY
491    TX
492    CA
493    ID
494    MI
495    CA
496    CA
497    OK
498    ME
499    AZ
500    MO
501    WA
502    NE
Name: state, Length: 503, dtype: object

#### Dsvid Gustafsson

##### Define
*Your definition here.*

##### Code

In [51]:
# Your cleaning code here
patients_clean.given_name = patients_clean.given_name.replace('Dsvid' , 'David')

##### Test

In [52]:
# Your testing code here
patients_clean[patients_clean.surname=='Gustafsson']

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

   phone_number                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
*Your definition here. Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) for one method used in solution, [documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) for one function used in the solution, and [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) for another method used in the solution.*

* change the data types of assigned sex and state from object to category
* change the data types of birthdate from object to date_time
* two columns dose_start and dose_end should be integer
* replace "u" with "" for starting and ending doses columns

##### Code

In [53]:
# Your cleaning code here
#patients_clean.dtypes
#treatments.dtypes
patients_clean[['assigned_sex','state','zip_code']] = patients_clean[['assigned_sex','state','zip_code']].apply(lambda x: x.astype('category'))
patients_clean['birthdate'] = pd.to_datetime( patients_clean['birthdate'] )

# to use strip(), need to convert to string first
treatments_clean['dose_start'] = treatments_clean['dose_start'].str.strip('u').astype(int)
treatments_clean['dose_end' ] = treatments_clean['dose_end' ].str.strip('u').astype(int)

# to use replace, need to be data frame 
#treatments_clean['dose_start'] = treatments_clean['dose_start'].replace('u','').astype(int)
#treatments_clean['dose_start'] = treatments_clean['dose_start'].replace('u','').astype(int)




##### Test

In [54]:
# Your testing code here
patients_clean.dtypes

patient_id               int64
assigned_sex          category
given_name              object
surname                 object
address                 object
city                    object
state                 category
zip_code              category
country                 object
birthdate       datetime64[ns]
weight                 float64
height                   int64
bmi                    float64
phone_number            object
email                   object
dtype: object

In [55]:
treatments_clean[['dose_start' ,'dose_end' ]] 

dose_start  dose_end
0            41        48
1            33        36
2            37        42
3            31        38
4            30        36
5            29        36
6            29        36
7            29        38
8            27        37
9            55        68
10           28        37
11           29        39
12           53        60
13           31        41
14           42        51
15           42        49
16           35        39
17           47        58
18           45        48
19           24        37
20           44        55
21           37        43
22           43        47
23           36        42
24           29        37
25           57        64
26           54        67
27           34        42
28           29        37
29           61        67
..          ...       ...
319          37        38
320          28        24
321          24        26
322          20        21
323          46        46
324          26        27
325          34        33
326          30        35
327          33        33
328          35        34
329          30        27
330          37        41
331          43        40
332          34        34
333          40        36
334          37        40
335          26        23
336          40        45
337          52        52
338          37        39
339          42        42
340          26        25
341          36        33
342          42        41
343          32        33
344          55        51
345          26        23
346          22        23
347          28        26
348          42        44

[349 rows x 2 columns]

#### Multiple phone number formats

##### Define
*Your definition here. Hint: helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).*
* Strip all " ", "-", "(", ")", and "+" and store each number without any formatting.
* Pad the phone number with a 1 if the length of the number is 10 digits (we want country code).

##### Code

In [56]:
# Your cleaning code here
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '').str.pad(11, fillchar='1')
# '\D+' stands for all string except digits

##### Test

In [57]:
# Your testing code here
patients_clean.phone_number

0      19517199170
1      12175693204
2      14023636804
3      17326368246
4      13345157487
5      13863345237
6      17755335933
7      14087783236
8      18162659578
9      17187959124
10     19064788949
11     19312070839
12     14345092614
13     12054178095
14     19187123469
15     12762251955
16     15597657836
17     16788298578
18     14067752696
19     15305328397
20     17814471763
21     19734455341
22     19147456108
23     18133559476
24     18458587707
25     15058284955
26     13096718852
27     19858147603
28     12678957462
29     18458587707
          ...     
473    15089052371
474    16782633564
475    16509627179
476    12679723749
477    18432126421
478    14157556435
479    19094582515
480    12103265509
481    19187062776
482    13616934960
483    17315770292
484    13366972005
485    18506590417
486    12546814504
487    17852291188
488    13524534601
489    19252835425
490    19146369304
491    12102228684
492    17144962264
493    12083881065
494    15867

#### Default John Doe data

##### Define
*Your definition here. Recall that it is assumed that the data that this John Doe data displaced is not recoverable.*
* Remove the non-recoverable John Doe records from the patients table

##### Code

In [71]:
patients_clean = patients_clean[patients_clean.surname != 'Doe']

##### Test

In [59]:
# Your testing code here
patients_clean.surname.value_counts()

Jakobsen        3
Taylor          3
Batukayev       2
Silva           2
Lâm             2
Souza           2
Collins         2
Correia         2
Lương           2
Tucker          2
Bùi             2
Gersten         2
Tạ              2
Woźniak         2
Grímsdóttir     2
Ogochukwu       2
Schiavone       2
Berg            2
Cindrić         2
Kadyrov         2
Liễu            2
Lund            2
Hueber          2
Dratchev        2
Nilsen          2
Aranda          2
Cabrera         2
Johnson         2
Parker          2
Kowalczyk       2
               ..
Rice            1
Hervinsson      1
Mikkelsen       1
Mathiesen       1
Thạch           1
Day             1
Mata            1
Štěpánek        1
Aličajić        1
Wiśniewska      1
Rokavc          1
Kašná           1
Azevedo         1
Ščančar         1
Efrem           1
Richardson      1
Nnonso          1
Adams           1
Jakšić          1
Stanić          1
House           1
Frederiksen     1
Guðbrandsson    1
Reynoso         1
Sokołowska

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
*Your definition here.*
* remove Jake Jakobsen, Pat Gersten, and Sandy Taylor rows from the patients table. They are all nicknames, second occurence of 
duplicate. 


##### Code

In [67]:
sum(patients_clean.address.duplicated())

11

In [69]:
sum(patients_clean.address.notnull())

482

In [60]:
# Your cleaning code here   duplicated but not null values 
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

##### Test

In [61]:
# Your testing code here
patients_clean[ patients_clean.surname == 'Jakobsen' ]

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
432      Houston    TX    77020  United States 1962-11-25   185.2      67   

      bmi phone_number                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

In [62]:
patients_clean[ patients_clean.surname == 'Gersten' ]

patient_id assigned_sex given_name  surname            address  city  \
97          98         male    Patrick  Gersten  2778 North Avenue  Burr   

   state zip_code        country  birthdate  weight  height   bmi  \
97    NE    68324  United States 1954-05-03   138.2      71  19.3   

   phone_number                     email  
97  14028484923  PatrickGersten@rhyta.com

In [63]:
patients_clean[ patients_clean.surname == 'Taylor' ]

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

    state zip_code        country  birthdate  weight  height   bmi  \
131    WV    25962  United States 1960-10-23   206.1      64  35.4   
426    FL    33179  United States 1992-09-02   186.6      69  27.6   

    phone_number                       email  
131  13044382648    SandraCTaylor@dayrep.com  
426  13054346299  RogelioJTaylor@teleworm.us

#### kgs instead of lbs for Zaitseva weight

##### Define
*Use advanced indexing to isolate the row where the surname is Zaitseva and convert the entry in its weight field from kg to lbs

##### Code

In [64]:
# Your cleaning code here
weight_kg = patients_clean.weight.min()
mask = patients_clean.surname == 'Zaitseva'
column_name = 'weight'
patients_clean.loc[mask, column_name] = weight_kg * 2.20462

#  To replace multiples columns convert to numpy array using .values:
#  df.loc[df.A==0, ['B', 'C']] = df.loc[df.A==0, ['B', 'C']].values / 2    
# df.A==0 is condition , once meet the condition,  column B , C values /2



##### Test

In [65]:
# Your testing code here
patients_clean[patients_clean.surname == 'Zaitseva']

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country  birthdate      weight  height  \
210  Wooster    OH    44691  United States 1938-11-26  107.585456      63   

      bmi phone_number                          email  
210  19.1  13302022145  CamillaZaitseva@superrito.com